### ISSUE TITLE
Failed to create CUDAExecutionProvider 

### ISSUE THREAD

**Issue by Apisteftos** [microsoft/onnxruntime#16128]
### Describe the issue

I am trying to inference with the GPU, while the CUDA - GPU is detected in the runtime fails to create a CUDAExecutionProvider.

### To reproduce

import onnxruntime as rt
print(rt.get_device())

`GPU` 

from the [API](https://onnxruntime.ai/docs/api/python/api_summary.html). My code is almost the same as in API 

X = (1, 3, 224, 224) from numpy array. Exactly at X_ortvalue an error occurs:

X_ortvalue = onnxruntime.OrtValue.ortvalue_from_numpy(X, 'cuda', 0)
session = onnxruntime.InferenceSession(
        'model.onnx',
        providers=['CUDAExecutionProvider', 'CPUExecutionProvider'])
)
io_binding = session.io_binding()
io_binding.bind_input(
        name='input',
        device_type='cuda',
        device_id=0,
        element_type=np.float32,
        shape=X_ortvalue.shape(),
        buffer_ptr=X_ortvalue.data_ptr()
)

io_binding.bind_output(
        name='output',
        device_type='cuda',
        device_id=0
)

io_binding.bind_output('output', 'cuda')
session.run_with_iobinding(io_binding)
ort_output = io_binding.get_outputs()[0] 

Error:
--------
2023-05-27 00:53:18.055404136 [W:onnxruntime:Default, onnxruntime_pybind_state.cc:541 CreateExecutionProviderInstance] Failed to create CUDAExecutionProvider. Please reference https://onnxruntime.ai/docs/reference/execution-providers/CUDA-ExecutionProvider.html#requirements to ensure all dependencies are met.
Backend QtAgg is interactive backend. Turning interactive mode on.
Inconsistency detected by ld.so: dl-version.c: 204: _dl_check_map_versions: Assertion `needed != NULL' failed!

 

### Platform

Linux

### OS Version

Ubuntu 22.04

### ONNX Runtime Installation

pip installation

### ONNX Runtime Version or Commit ID

1.14

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

CUDA

### Execution Provider Library Version

CUDA 11.6

**Comment by Apisteftos** [comment]
I don't know how, I just imported torch and the problem solved. I am not using at all pytorch, but it seems activates 'CUDA' libraries which they are missing. Does anyone has any idea what causes this issue?

**Comment by mizhitian-xiaomi** [comment]
Did you solve it? 
I'm also having the same issue in docker containers

**Comment by Apisteftos** [comment]
I just imported the torch. Import in your python script: import torch try in the main or in the current script and it will activate the CUDA libs. Do not ask me how is that happening, I don't know but you can find here more information. #13264 and here #11092

**Comment by kopyl** [comment]
Importing Torch helps. But we need a better solution since Torch takes a lot of time to load.

**Comment by mizhitian-xiaomi** [comment]
thank u. 
I solved the problem. I checked the `/dev` dir ，then didn't found `nvidia-uvm` module. I run one sample that nvidia driver provided and it installed the nvidia-uvm automatically. then i can call onnxruntime-gpu in docker containers.

use `lsmod | grep -i nvidia` to check if there is an `nvidia-uvm` module in the container.

**Comment by Apisteftos** [comment]
You mean you show the nvidia_uvm flag = 0? 

```
nvidia_uvm           1400832  0
nvidia_drm             73728  9
nvidia_modeset       1241088  14 nvidia_drm
nvidia              56328192  838 nvidia_uvm,nvidia_modeset
drm_kms_helper        200704  1 nvidia_drm
drm                   581632  13 drm_kms_helper,nvidia,nvidia_drm

```


I am having again issues since I uninstalled onnxruntime and onnxruntime-gpu

**Comment by Apisteftos** [comment]
Found the problem: 
In python packages they are missing the pre_reloaded cuda libraries:

According to (ONNXRUNTIME)[https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/_ld_preload.py] these libraries can be modified: That means the libraries are missing from the policies or the onnxruntime, Also `__init__` file is totally empty and I am not sure is it should be. 


```
# -------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.
# --------------------------------------------------------------------------


# This file can be modified by setup.py when building a manylinux2010 wheel
# When modified, it will preload some libraries needed for the python C extension

```

I this file should be the cuda libraries. I was working today on a server and doing some research on onnxruntime and here were the loaded libraries such as:

```
        libcudart.so: CUDA runtime library
	libcublas.so: CUDA Basic Linear Algebra Subprograms (BLAS) library
	libcusparse.so: CUDA sparse matrix library
	libcurand.so: CUDA random number generation library
	libcufft.so: CUDA Fast Fourier Transform (FFT) library
	libcusolver.so: CUDA linear solver library
	libnvrtc.so: NVIDIA Runtime Compilation (NVRTC) library (for compiling CUDA code at runtime)

```

To find which cuda libraries are installed in the machine we need:


```
        find  -name "libcudart.so*"
	find  -name "libcublas.so*"
	find  -name "libcusparse.so*"
	find  -name "libcurand.so*"
	find  -name "libcufft.so*"
	find  -name "libcusolver.so*"
	find  -name "libnvrtc.so*"

```

**Comment by Apisteftos** [comment]
Now it works without importing the pytorch: 

I am tried to build it from the source and it had several problems.
I uninstalled onnxruntime and onnxruntime-gpu and installed again

Firstly I copied inside the venv/lib64/python3.10/site-packages/onnxruntime/capi/_ld_preload.py  which was empty.

```
from ctypes import CDLL, RTLD_GLOBAL
try:
    _libcublasLt = CDLL("libcublasLt.so.11", mode=RTLD_GLOBAL)
    _libcublas = CDLL("libcublas.so.11", mode=RTLD_GLOBAL)
    _libcudnn = CDLL("libcudnn.so.8", mode=RTLD_GLOBAL)
    _libcurand = CDLL("libcurand.so.10", mode=RTLD_GLOBAL)
    _libcufft = CDLL("libcufft.so.10", mode=RTLD_GLOBAL)
    _libcudart = CDLL("libcudart.so.11.0", mode=RTLD_GLOBAL)
except OSError:
    import os
    os.environ["ORT_CUDA_UNAVAILABLE"] = "1"
from ctypes import CDLL, RTLD_GLOBAL
try:
    _libcudnn = CDLL("libcudnn.so.8", mode=RTLD_GLOBAL)
    _libcublas = CDLL("libcublas.so.11", mode=RTLD_GLOBAL)
    _libnvinfer = CDLL("libnvinfer.so.8", mode=RTLD_GLOBAL)
    _libnvinfer_plugin = CDLL("libnvinfer_plugin.so.8", mode=RTLD_GLOBAL)
    _libnvonnxparser = CDLL("libnvonnxparser.so.8", mode=RTLD_GLOBAL)
    _libcudart = CDLL("libcudart.so.11.0", mode=RTLD_GLOBAL)
except OSError:
    import os
    os.environ["ORT_TENSORRT_UNAVAILABLE"] = "1"
```
but it didn't worked

Then I uninstalled the again onnxruntime and onnxruntime-gpu and tried to installed from source. The installation from source had several issues and it was really very slowly. As a was searching inside the packages I found a remtand of previous onnxruntime packages I just deleted in lib64 and lib the onnxruntime packages and then I installed again. 

After the re-installation I checked if the  `_ld_preload.py` file was empty, but luckily  all the libs installed correctly without any problem. I removed also the `import torch` from my scripts and works without that.

**Comment by snnn** [comment]
We made an important fix in the latest 1.16.0 release. With the new release you should not see the same error , "ld.so: dl-version.c xxxx" again. Please open a new issue if you still have problems on installing(or using) ONNX Runtime GPU packages. 

Thanks.

**Comment by Arcitec** [comment]
> I hate that this fixed it, but yes, importing `torch` before `onnxruntime` fixed it for me as well.

# Read this if you want to know why this issue happens:

The problem is that ONNX doesn't know how to search for CUDA. PyTorch knows how to search for it, and adds it to Python's internal path, so that ONNX can later find it.

The bug/issue is with ONNX library. I have coded a workaround here:

https://github.com/cubiq/ComfyUI_IPAdapter_plus/issues/238nn

> I hate that this fixed it, but yes, importing `torch` before `onnxruntime` fixed it for me as well.

# Read this if you want to know why this issue happens:

The problem is that ONNX doesn't know how to search for CUDA. PyTorch knows how to search for it, and adds it to Python's internal path, so that ONNX can later find it.

The bug/issue is with ONNX library. I have coded a workaround here:

https://github.com/cubiq/ComfyUI_IPAdapter_plus/issues/238

PS: @snnn If you could get this information to the team, it would be fantastic if future ONNX versions can automatically find the CUDA libraries in the same way that PyTorch does it. :)
